src/util/time.rs

```diff
@@ -247,44 +237,74 @@ impl Instant {
         match *self {
             Instant::Monotonic(t) => {
                 let now = monotonic_now();
-                elapsed_duration(now, t)
+                Instant::elapsed_duration(now, t)
             }
             Instant::MonotonicCoarse(t) => {
                 let now = monotonic_coarse_now();
-                elapsed_duration(now, t)
+                Instant::elapsed_duration_coarse(now, t)
             }
         }
     }
 
     pub fn duration_since(&self, earlier: Instant) -> Duration {
-        let later = self.get_timespec();
-        let earlier = earlier.get_timespec();
-        elapsed_duration(later, earlier)
+        match (*self, earlier) {
+            (Instant::Monotonic(later), Instant::Monotonic(earlier)) => {
+                Instant::elapsed_duration(later, earlier)
+            }
+            (Instant::MonotonicCoarse(later), Instant::MonotonicCoarse(earlier)) => {
+                Instant::elapsed_duration_coarse(later, earlier)
+            }
+            _ => {
+                panic!("duration between different types of Instants");
+            }
+        }
     }
 
-    fn get_timespec(&self) -> Timespec {
-        match *self {
-            Instant::Monotonic(t) |
-            Instant::MonotonicCoarse(t) => t,
+    fn elapsed_duration(later: Timespec, earlier: Timespec) -> Duration {
+        if later >= earlier {
+            (later - earlier).to_std().unwrap()
+        } else {
+            panic!("system time jumped back, {:.9} -> {:.9}",
+                   earlier.sec as f64 + earlier.nsec as f64 / NANOSECONDS_PER_SECOND as f64,
+                   later.sec as f64 + later.nsec as f64 / NANOSECONDS_PER_SECOND as f64);
         }
     }
-}
 
-impl PartialEq for Instant {
-    fn eq(&self, other: &Instant) -> bool {
-        self.get_timespec().eq(&other.get_timespec())
+    fn elapsed_duration_coarse(later: Timespec, earlier: Timespec) -> Duration {
+        let later_ms = later.sec * MILLISECOND_PER_SECOND +
+                       (later.nsec / NANOSECONDS_PER_MILLISECOND) as i64;
+        let earlier_ms = earlier.sec * MILLISECOND_PER_SECOND +
+                         (earlier.nsec / NANOSECONDS_PER_MILLISECOND) as i64;
+        if later_ms >= earlier_ms {
+            Duration::from_millis((later_ms - earlier_ms) as u64)
+        } else {
+            panic!("system time jumped back, {:.3} -> {:.3}",
+                   earlier.sec as f64 + earlier.nsec as f64 / NANOSECONDS_PER_SECOND as f64,
+                   later.sec as f64 + later.nsec as f64 / NANOSECONDS_PER_SECOND as f64);
+        }
     }
 }
 
-impl Ord for Instant {
-    fn cmp(&self, other: &Instant) -> Ordering {
-        self.get_timespec().cmp(&other.get_timespec())
+impl PartialEq for Instant {
+    fn eq(&self, other: &Instant) -> bool {
+        match (*self, *other) {
+            (Instant::Monotonic(this), Instant::Monotonic(other)) |
```

Comment:
It's already covered. 
https://github.com/pingcap/tikv/pull/2165/files/6dfd8f12f8be1be1e6450147716333cb7df9ff2b#diff-5a3b02da8619a0100baffd3477a68b21R411